src/deepagents/backends/protocol.py

```diff
@@ -1,218 +1,200 @@
-            file_path: Absolute file path (e.g., "/notes.txt", "/memories/agent.md")
-            content: File content as a string
+            pattern: Search pattern (implementation-specific: literal or regex).
+            path: Directory to search in (default "/").
+            glob: Optional glob pattern to filter files (e.g., "*.py").
+            output_mode: Output format - "files_with_matches" (file paths), "content"
+                (matching lines with context), or "count" (match counts per file).
 
         Returns:
-            - Command object for StateBackend (uses_state=True) to update LangGraph state
-            - Success message string for other backends, or error if file already exists
+            Formatted search results based on output_mode, or message if no matches found.
 
-        Error cases:
-            - Returns error message if file already exists (should use edit instead)
+        Note:
+            This is a basic implementation. Missing features: context lines (-A/-B/-C),
+            line numbers (-n), case sensitivity (-i), output limiting, file type filters,
+            multiline support.
         """
         ...
 
-    def edit(
-            self,
-            file_path: str,
-            old_string: str,
-            new_string: str,
-            replace_all: bool = False,
-    ) -> Command | str:
-        """Edit a file by replacing string occurrences.
+    @abstractmethod
+    def glob(self, pattern: str, path: str = "/") -> list[str]:
+        """Find files matching a glob pattern.
 
         Args:
-            file_path: Absolute file path (e.g., "/notes.txt", "/memories/agent.md")
-            old_string: String to find and replace
-            new_string: Replacement string
-            replace_all: If True, replace all occurrences; if False, require unique match
+            pattern: Glob pattern (e.g., "**/*.py", "*.txt", "/subdir/**/*.md").
+            path: Base directory to search from (default "/").
 
         Returns:
-            - Command object for StateBackend (uses_state=True) to update LangGraph state
-            - Success message string for other backends, or error message on failure
-
-        Error cases:
-            - "Error: File '{file_path}' not found" if file doesn't exist
-            - "Error: String not found in file: '{old_string}'" if string not found
-            - "Error: String '{old_string}' appears {n} times. Use replace_all=True..."
-              if multiple matches found and replace_all=False
+            List of absolute file paths matching pattern. Empty list if no matches.
         """
         ...
 
-@runtime_checkable
-class StateBackendProvider(Protocol):
 
-    def get_backend(self, runtime: ToolRuntime) -> StateBackendProtocol:
-        """Get the backend."""
-        ...
\ No newline at end of file
+# Backend factory function types
+BackendProvider = Callable[[ToolRuntime], Backend]
+AsyncBackendProvider = Callable[[ToolRuntime], Awaitable[Backend]]
```

Comment:
We'll need async support at some point for all operations + the factory itself. Would suggest **not** adding it until we're ready to add full async support